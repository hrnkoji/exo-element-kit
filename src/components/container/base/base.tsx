import * as React from 'react';
import { State } from 'domain/store/main';
import { onClickGetQuote } from 'domain/middleware/user';
import Button from '@material-ui/core/Button';
import Replay from '@material-ui/icons/Replay';
import Face from '@material-ui/icons/Face';
import FormatQuote from '@material-ui/icons/FormatQuote';
import Card from '@material-ui/core/Card';
import CardActions from '@material-ui/core/CardActions';
import CardContent from '@material-ui/core/CardContent';
import Typography from '@material-ui/core/Typography';
import styled from 'styled-components';

const StyledButton = styled(Button)`
  && {
    background-color: black;
    border-radius: 50px;
    border: 0;
    color: white;
    height: 48px;
    padding: 0 30px;
  }
`;

type BaseProps = {
  state: State;
};
export function Base({ state }: BaseProps) {
  return (
    <div>
      <Card>
        <CardContent>
          <Typography variant="headline" component="h3">
            {state.quote.quote}
          </Typography>
          <Typography component="p">{state.quote.author}</Typography>
        </CardContent>
        <CardActions>
          <StyledButton onClick={onClickGetQuote} size="small">
            Get Quote <FormatQuote />
          </StyledButton>
        </CardActions>
      </Card>
    </div>
  );
}
