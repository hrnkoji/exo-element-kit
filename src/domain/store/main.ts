import { createAtom } from 'js-atom';

export type Quote = {
  quote: string,
  author: string,
};

export type State = {
  quote: Quote
};

const defaultState: State = {
  quote: {
    quote: 'The wit makes fun of other persons the satirist makes fun of the world the humorist makes fun of himself, but in so doing, he identifies himself with people - that is, people everywhere, not for the purpose of taking them apart, but simply revealing their true nature',
    author: 'James Thurber',
  }
};

export const store = createAtom(defaultState);

export function state(): State {
  return store.deref();
}


