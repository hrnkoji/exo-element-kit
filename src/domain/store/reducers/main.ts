import { getLogger } from 'domain/logger';
import { store, State } from 'domain/store/main';
import { Quote } from 'domain/store/main';

const logger = getLogger('@store/reducers');

export function updateQuote(item: Quote) {
  logger.debug(`Update item ${item}`);
  return store.swap(oldState => ({ ...oldState, quote: item }));
}

export function updateCustomElementAttributes(attributes: any) {
  logger.debug(`Update custom attributes ${attributes}`);
  return store.swap(oldState => ({ ...oldState,  attributes: { ...attributes }}));
}