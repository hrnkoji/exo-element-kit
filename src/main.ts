import { store } from 'domain/store/main';
import render from 'renderer';
import { updateCustomElementAttributes } from 'domain/store/reducers/main';
import exoElementConfig from '../exo-element.config';
import { map, each } from 'lodash';

class ExoElement extends HTMLElement {

  static get observedAttributes() {
    return map(exoElementConfig.attributes, 'name');
  }

  constructor() {
    super();
    render();
    this.render();
  }

  render () {
    store.addWatch('renderLoop', render);
    let obj = {};
    updateCustomElementAttributes(
      each(exoElementConfig.attributes,
           function(attr) { obj[attr.name] = attr.value }));
  }

}

window.customElements.define(exoElementConfig.name, ExoElement);
