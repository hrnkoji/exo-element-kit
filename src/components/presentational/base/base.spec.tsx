import React from 'react';
import ReactDOM from 'react-dom';
import { create } from 'react-test-renderer';
import { Base } from './base';
import { state } from 'domain/store/main';

test('Renders correctly with default data', () => {
  const tree1 = create(<Base state={state()} />).toJSON();
  expect(tree1).toMatchSnapshot();
});
