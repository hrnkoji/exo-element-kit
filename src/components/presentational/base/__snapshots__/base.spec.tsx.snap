// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly with default data 1`] = `
<div>
  <div
    className="MuiPaper-root-2 MuiPaper-elevation2-6 MuiPaper-rounded-3 MuiCard-root-1"
  >
    <div
      className="MuiCardContent-root-29"
    >
      <h3
        className="MuiTypography-root-30 MuiTypography-headline-35"
      >
        The wit makes fun of other persons the satirist makes fun of the world the humorist makes fun of himself, but in so doing, he identifies himself with people - that is, people everywhere, not for the purpose of taking them apart, but simply revealing their true nature
      </h3>
      <p
        className="MuiTypography-root-30 MuiTypography-body1-39"
      >
        James Thurber
      </p>
    </div>
    <div
      className="MuiCardActions-root-54"
    >
      <button
        className="MuiButtonBase-root-72 MuiButton-root-56 MuiButton-sizeSmall-69 MuiCardActions-action-55 sc-bdVaJa bHCexJ"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="0"
        type="button"
      >
        <span
          className="MuiButton-label-57"
        >
          Get Quote 
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-75"
            color={undefined}
            focusable="false"
            viewBox="0 0 24 24"
          >
            <g>
              <path
                d="M6 17h3l2-4V7H5v6h3zm8 0h3l2-4V7h-6v6h3z"
              />
            </g>
          </svg>
        </span>
        <span
          className="MuiTouchRipple-root-81"
        />
      </button>
    </div>
  </div>
</div>
`;
